### DISCUSSION TITLE
doc for freebsd13.1 and flaresolverr3.1.2

### DISCUSSION THREAD

**Discussion by furonair** [FlareSolverr/FlareSolverr#748] [Show and tell]
Here are the instructions in order to get **FlareSolverr** configured and running at boot in a **FreeBSD** system or in a **FreeBSD** jail (which is my case).

> This was made in freebsd 13.1-RELEASE-p6 in jail.

> All instructions consider that you will be using the **root** user to execute the commands and flaresover will run **flaresolver** with flaresolver user. Some caveat from the script for to do some **particular** things.

I'll not enter in details about creation of jails.
But this should work to on the FreeBSD host OS.

## 1. Check if you need to set your date/time

```
rm /etc/localtime
ln -s /usr/share/zoneinfo/America/Sao_Paulo /etc/localtime
```

## 2. Update and upgrade your packages

Before anything else, make sure that your packages are updated and upgraded:

```
pkg update
pkg upgrade
```

## 3. Get FlareSolverr

Here you have the option to use **Git** or download the sources directly. It doesn't matter what you choose, the final product will be the same. 

The advantage of the **Git** way is that it is easier to keep **FlareSolverr** updated.

So, follow 3.1 or 3.2, as you wish.

### 3.1 Using Git

Install the git-tiny package:
```
pkg install git-tiny
```

Then go to /usr/local/share folder and clone the repository:
```
cd /tmp/
git clone https://github.com/FlareSolverr/FlareSolverr.git
```

### 3.2 Using Zip

Go to the [**FlareSolverr** Github page](https://github.com/FlareSolverr/FlareSolverr), click on **Code**, then right-click on "**Download ZIP**" and then choose "**Copy link address**":

![image](https://user-images.githubusercontent.com/6900313/147861613-2e698c32-3672-41e2-ab04-34c5ca11fa69.png)

Then execute the **fetch** command to download the zip file, by pasting the copied address in the command line, and then run the **unzip** command to extract it:

```
cd /tmp/
fetch https://github.com/FlareSolverr/FlareSolverr/archive/refs/heads/master.zip
unzip master.zip
```

## 4. Installing the necessary packages and load necessary kernel module

Once you have the FlareSolverr source code downloaded through Git or Zip in /usr/local/share/FlareSolverr, the next step is to download the necessary packages.

### 4.1 Load Kernel module
Load kernel module required by linux compatibility package (probably just linux64 is required but is not my case and I did test). If you do please confirm.

```
#to load kernels module now
kldload -n -v linux
kldload -n -v linux64

#to load kernels module at boot
sysrc kld_list+="linux"
sysrc kld_list+="linux64"
```

### 4.2 Install package
Install these packages and all their required dependencies:

```
pkg install python39 py39-pip chromium xorg-vfbserver linux_base-c7
```

> 158 of packages aprox should be installed and will require near a 3GiB of storage.

## 5. Installing virtuaenv and requirement

To keep most of thing clean on the host after the build will be completed

### 5.1. virtual
```
pip install virtualenv
```

### 5.2. Create and activate virtual env
```
virtualenv /tmp/venv
source /tmp/venv/bin/activate.csh
```

### 5.3. Installing requirements
```
cd /tmp/FlareSolverr-master
pip install pyinstaller
pip install -r requirements.txt
pip install -r test-requirements.txt
```

## 6. building the apps

```
cd ./src/
python3 build_package.py
cd ..
```

### 7.1. Special folders
Thoses things are requireds because the program ask for access to particular folder.

```
mkdir /app
chmod 777 /app
mkdir /.undetected_chromedriver
chmod 755 /.undetected_chromedriver
mkdir /root/.undetected_chromedriver
chmod 755 /root/.undetected_chromedriver
cp /usr/local/bin/chromedriver /app/chromedriver
chmod 777 /app/chromedriver
cp /usr/local/bin/chromedriver /app/chromedriver.exe
chmod 777 /app/chromedriver.exe
```

### 7.3. Linux compatibility
Thoses things ar required for linux compatibility.

```
rm -rf ./dist/package/flaresolverr/chrome/*
cp /usr/local/bin/chrome ./dist/package/flaresolverr/chrome/chrome

#showing the ELF mark on the file
brandelf ./dist/package/flaresolverr/*.so.*

#change the ELF mark on the file (if is not 'Linux')
brandelf -t Linux ./dist/package/flaresolverr/*.so.*

#showing the result of the change
brandelf ./dist/package/flaresolverr/*.so.*


# seams to be a a bug of the linux compatibility package installed earlier the file did not exist, created the file with
# a simlink with a existing simlink in the hope it will be easier to find than the real file

ln -s /compat/linux/usr/lib64/libc.so.6 /compat/linux/usr/lib64/libc.so.7
```

## 8. Create user
```
pw useradd flaresolverr -s /usr/sbin/nologin
```

## 9. Moving folder and changing owner and permission
```
mv ./dist/package/flaresolverr /usr/local/share/flaresolverr
chown -R flaresolverr:flaresolverr /usr/local/share/flaresolverr
chmod -R 770 /usr/local/share/flaresolverr
```

## 10. Create the rc.d config file to define the service

create the file
```
touch /usr/local/etc/rc.d/flaresolverr
chmod 755 /usr/local/etc/rc.d/flaresolverr
```

edit the file with your prefered text editor
```
#!/bin/sh
#
# PROVIDE: flaresolverr
# REQUIRE: networking
# KEYWORD: shutdown

# flaresolverr_enable="YES"

. /etc/rc.subr

name="flaresolverr"
rcvar="flaresolverr_enable"


flaresolverr_log_file="/var/log/flaresolverr/flaresolverr.log"

start_precmd=flaresolverr_prestart
stop_postcmd=flaresolverr_poststop

load_rc_config ${name}
# Defaults
# Usable parameter in rc.conf with those default parameter
: ${flaresolverr_enable:=NO}
: ${flaresolverr_deamon_user:=flaresolverr}
: ${flaresolverr_path:="/usr/local/share/flaresolverr"}
: ${flaresolverr_LOG_LEVEL:="info"}
: ${flaresolverr_LOG_HTML:="false"}
: ${flaresolverr_CAPTCHA_SOLVER:="false"}
: ${flaresolverr_TZ:="UTC"}
: ${flaresolverr_HEADLESS:="true"}
: ${flaresolverr_BROWSER_TIMEOUT:="40000"}
: ${flaresolverr_PORT:="8191"}
: ${flaresolverr_HOST:="0.0.0.0"}
: ${flaresolverr_TEST_URL:="https://www.google.com"}

pidfile_deamon="/var/run/flaresolverr/flaresolverr_deamon.pid"
pidfile_child="/var/run/flaresolverr/flaresolverr_child.pid"
procname="${flaresolverr_path}/flaresolverr"
command="/usr/sbin/daemon"

command_args="-f -p ${pidfile_child} -P ${pidfile_deamon} -o $flaresolverr_log_file -u ${flaresolverr_deamon_user} ${flaresolverr_path}/flaresolverr"

flaresolverr_prestart()
{

    install -d -o root -g $flaresolverr_deamon_user /var/run/flaresolverr
    install -d -o root -g $flaresolverr_deamon_user /var/log/flaresolverr

    export LOG_LEVEL=$flaresolverr_LOG_LEVEL
    export LOG_HTML=$flaresolverr_LOG_HTML
    export CAPTCHA_SOLVER=$flaresolverr_CAPTCHA_SOLVER
    export TZ=$flaresolverr_TZ
    export HEADLESS=$flaresolverr_HEADLESS
    export BROWSER_TIMEOUT=$flaresolverr_BROWSER_TIMEOUT
    export PORT=$flaresolverr_PORT
    export HOST=$flaresolverr_HOST
    export TEST_URL=$flaresolverr_TEST_URL

    
}

flaresolverr_poststop()
{
    sleep 2
    killall -u ${flaresolverr_deamon_user} -c Xvfb
}

run_rc_command "$1"

```

## 11. Clean no more usefull files and folders
rm -rf /tmp/venv
rm -rf /tmp/FlareSolverr-master
rm -rf /tmp/build
rm -rf /tmp/master.zip

## 12. Enable service in rc.conf and start the deamon
```
sysrc flaresolverr_enable="YES"
```

You could find all allowed parameter to be used with the rc.d script within the file and configure the parameter with sysrc
Those parameter will be loaded as variable env

ex:
````
sysrc flaresolverr_PORT="80"
````

Now you should be able to start the deamon and enjoye!

```
service flaresolverr start

service flaresolverr status

tail -f /var/log/flaresolverr/flaresolverr.log
```

**Comment by ngosang** [comment]
Could you review and update this page in the Wiki?
https://github.com/FlareSolverr/FlareSolverr/wiki/Installing-FlareSolverr-in-FreeBSD-or-FreeNAS-TrueNAS-jail

**Reply by furonair** [reply]
Since is done for version 2.x in with node JS this page is not relevant anymore. I didn't have the right to do change or propose change in the wiki to add the information I added in this discution. I will  be glade to add this to the wiki for reference.

**Reply by ngosang** [reply]
I changed the Wiki. Thank you!

**Comment by furonair** [comment]
updated to follow the change in 3.1.2

**Reply by ilike2burnthing** [reply]
Thanks, updated.

**Comment by FroggMaster** [comment]
@ilike2burnthing Would you be able to review the information that I posted here:  https://github.com/FlareSolverr/FlareSolverr/issues/95#issuecomment-1793351556 and update the Wiki accordingly? I ran into a few issues when following the the guide and hope that the additional info helps anyone else that might find it helpful. Thanks. :)

**Reply by dbolser** [reply]
Thanks @FroggMaster, did this get updated, or should I follow the guide here + your guide? (Or the guide on the wiki?) Where is the definitive guide :-)

Cheers,

**Reply by ilike2burnthing** [reply]
Click the linked issue and scroll down.

**Reply by dbolser** [reply]
So sorry, which link? It says 'LEGACY' is this whole guide old?

I'm seeing this error:

```
(venv) root@flaresolverr:/tmp/FlareSolverr # tail -f /var/log/flaresolverr/flaresolverr.log
2025-01-25 18:04:55 INFO     FlareSolverr 3.3.21
2025-01-25 18:04:55 INFO     Testing web browser installation...
2025-01-25 18:04:55 INFO     Platform: FreeBSD-14.2-RELEASE-amd64-64bit-ELF
2025-01-25 18:04:55 INFO     Chrome / Chromium path: /usr/local/share/flaresolverr/_internal/chrome/chrome
/usr/local/share/flaresolverr/_internal/chrome/chrome: error while loading shared libraries: libnss3.so: cannot open shared object file: No such file or directory
2025-01-25 18:04:55 ERROR    Chrome / Chromium version not detected!
```

I know I'm being stupid, but I can normally follow a guide... don't know what I'm doing wrong.

**Reply by ilike2burnthing** [reply]
The only link in the comment you're replying to.

To skip a few steps here - https://github.com/FlareSolverr/FlareSolverr#from-source-code-freebsdtruenas-core

**Comment by dbolser** [comment]
Thanks @furonair, is the key information here getting all the versions lined up in a 'working' state, or is it around setting up the permissions / users 'just right'? I did the latter but still got issues, which I suspect may be related to versions (see https://github.com/FlareSolverr/FlareSolverr/discussions/1434).

If versions are the sticking point, I'll try to create a 'matrix' of python / FlareSolverr / FreeBSD where it works vs. fails.

**Comment by dbolser** [comment]
Right... Those instructions aren't working for me unfortunately.

**Reply by ilike2burnthing** [reply]
Given the error, did you run this?

> pkg install nss

**Reply by dbolser** [reply]
Thanks so much again @ilike2burnthing, in fact the above error was related to the out of date node version (I was following the old guide by mistake). On the new guide, on the forked code, it works :-) 🎉

Sorry for all the noise I've generated on the issues / discussion.

For absolute clarity (for idiots like me):

WORKING VERSION:
* FreeBSD RELEASE 14.2 (Jail)
  * Followed the guide here: https://github.com/FlareSolverr/FlareSolverr#from-source-code-freebsdtruenas-core
* FlareSolver version from here:
  * https://github.com/21hsmw/FlareSolverr/tree/nodriver-support

I needed to edit the code to point to chrome, but then it works.
